pkg/appgw/frontend_listeners.go

```diff
@@ -1,16 +1,97 @@
+
+	var httpListeners []n.ApplicationGatewayHTTPListener
+
+	for listener, config := range builder.getListenerConfigs(ingressList) {
+		httpListener := builder.newHTTPListener(listener, config.Protocol)
+		if config.Protocol == n.HTTPS {
+			sslCertificateID := builder.appGwIdentifier.sslCertificateID(config.Secret.secretFullName())
+			httpListener.SslCertificate = resourceRef(sslCertificateID)
+		}
+		httpListeners = append(httpListeners, httpListener)
+		legacyMap[listener] = &httpListener
+	}
+
+	// TODO(draychev): The second map we return is for compatibility w/ RequestRoutingRules and should be removed ASAP
+	return &httpListeners, legacyMap
+}
+
+// getListenerConfigs creates an intermediary representation of the listener configs based on the passed list of ingresses
+func (builder *appGwConfigBuilder) getListenerConfigs(ingressList []*v1beta1.Ingress) map[frontendListenerIdentifier]frontendListenerAzureConfig {
+	allListeners := make(map[frontendListenerIdentifier]frontendListenerAzureConfig)
 	for _, ingress := range ingressList {
-		_, _, configMap := builder.processIngressRules(ingress)
-		for k, v := range configMap {
-			httpListenersAzureConfigMap[k] = v
+		_, azListenerConfigs := builder.processIngressRules(ingress)
+		for listenerID, azConfig := range azListenerConfigs {
+			allListeners[listenerID] = azConfig
+		}
+	}
+
+	// App Gateway must have at least one listener - the default one!
+	if len(allListeners) == 0 {
+		allListeners[defaultFrontendListenerIdentifier()] = frontendListenerAzureConfig{
+			// Default protocol
+			Protocol: n.HTTP,
 		}
 	}
 
-	return httpListenersAzureConfigMap
+	return allListeners
+}
+
+func (builder *appGwConfigBuilder) newHTTPListener(listener frontendListenerIdentifier, protocol n.ApplicationGatewayProtocol) n.ApplicationGatewayHTTPListener {
+	frontendPortName := generateFrontendPortName(listener.FrontendPort)
+	frontendPortID := builder.appGwIdentifier.frontendPortID(frontendPortName)
+
+	return n.ApplicationGatewayHTTPListener{
+		Etag: to.StringPtr("*"),
+		Name: to.StringPtr(generateHTTPListenerName(listener)),
+		ApplicationGatewayHTTPListenerPropertiesFormat: &n.ApplicationGatewayHTTPListenerPropertiesFormat{
+			// TODO: expose this to external configuration
+			FrontendIPConfiguration: resourceRef(*builder.getPublicIPID()),
+			FrontendPort:            resourceRef(frontendPortID),
+			Protocol:                protocol,
+			HostName:                &listener.HostName,
+		},
+	}
+}
+
+func (builder *appGwConfigBuilder) getPublicIPID() *string {
+	var publicIPID *string
+	jsonConfigs := make([]string, 0)
+	for _, ip := range *builder.appGwConfig.FrontendIPConfigurations {
+		// Collect the JSON IP configs for debug purposes.
+		if jsonConf, err := ip.MarshalJSON(); err != nil {
+			jsonConfigs = append(jsonConfigs, string(jsonConf))
+		} else {
+			glog.Error("Could not marshall IP configuration:", *ip.ID, err)
```

Comment:
We are not using/logging jsonConfigs anywhere, do we keep it ?

Reply:
We use `jsonConfigs` all the way down on https://github.com/Azure/application-gateway-kubernetes-ingress/pull/184/files/3f46a93800aa5fc800fcbd6a2168250c7aee8d18#diff-b08d8c4173cd5b924fa92404420f18deR90 and then eventually log at https://github.com/Azure/application-gateway-kubernetes-ingress/pull/184/files/3f46a93800aa5fc800fcbd6a2168250c7aee8d18#diff-b08d8c4173cd5b924fa92404420f18deR93

So we use this in the Fatal `HTTP Listener was not able to find a Public IP address for App Gateway. Available IPs: ...`

Reply:
@akshaysngupta You were onto something here. This is buggy -- fixed here: https://github.com/Azure/application-gateway-kubernetes-ingress/pull/197/files